crates/extensions/c8y_firmware_manager/src/actor.rs

```diff
@@ -0,0 +1,756 @@
+use async_trait::async_trait;
+use c8y_api::smartrest::message::collect_smartrest_messages;
+use c8y_api::smartrest::message::get_smartrest_template_id;
+use c8y_api::smartrest::smartrest_deserializer::SmartRestFirmwareRequest;
+use c8y_api::smartrest::smartrest_deserializer::SmartRestRequestGeneric;
+use c8y_api::smartrest::smartrest_serializer::TryIntoOperationStatusMessage;
+use c8y_api::smartrest::topic::C8yTopic;
+use c8y_http_proxy::credentials::JwtRetriever;
+use download::Auth;
+use mqtt_channel::Topic;
+use nanoid::nanoid;
+use sha256::digest;
+use sha256::try_digest;
+use std::collections::HashMap;
+use std::fs;
+use std::os::unix::fs as unix_fs;
+use std::path::Path;
+use std::path::PathBuf;
+use tedge_actors::fan_in_message_type;
+use tedge_actors::Actor;
+use tedge_actors::ChannelError;
+use tedge_actors::DynSender;
+use tedge_actors::LoggingReceiver;
+use tedge_actors::MessageReceiver;
+use tedge_actors::RuntimeError;
+use tedge_actors::RuntimeRequest;
+use tedge_actors::Sender;
+use tedge_actors::WrappedInput;
+use tedge_api::OperationStatus;
+use tedge_downloader_ext::DownloadRequest;
+use tedge_downloader_ext::DownloadResult;
+use tedge_mqtt_ext::MqttMessage;
+use tedge_timer_ext::SetTimeout;
+use tedge_timer_ext::Timeout;
+use tedge_utils::file::move_file;
+use tedge_utils::file::PermissionEntry;
+use tracing::error;
+use tracing::info;
+use tracing::warn;
+
+use crate::config::FirmwareManagerConfig;
+use crate::error::FirmwareManagementError;
+use crate::message::get_child_id_from_child_topic;
+use crate::message::DownloadFirmwareStatusMessage;
+use crate::message::FirmwareOperationRequest;
+use crate::message::FirmwareOperationResponse;
+use crate::operation::ActiveOperationState;
+use crate::operation::FirmwareOperationEntry;
+use crate::operation::OperationKey;
+
+pub type OperationSetTimeout = SetTimeout<OperationKey>;
+pub type OperationTimeout = Timeout<OperationKey>;
+
+pub type IdDownloadResult = (String, DownloadResult);
+pub type IdDownloadRequest = (String, DownloadRequest);
+
+fan_in_message_type!(FirmwareInput[MqttMessage, OperationTimeout, IdDownloadResult] : Debug);
+fan_in_message_type!(FirmwareOutput[MqttMessage, OperationSetTimeout, IdDownloadRequest] : Debug);
+
+pub struct FirmwareManagerActor {
+    config: FirmwareManagerConfig,
+    active_child_ops: HashMap<OperationKey, ActiveOperationState>,
+    reqs_pending_download: HashMap<String, SmartRestFirmwareRequest>,
+    message_box: FirmwareManagerMessageBox,
+}
+
+impl FirmwareManagerActor {
```

Comment:
The code is clean with a regular structure. However, it's a bit difficult to understand the workflow?

Some function comments would help. Notably to tell when an even handler can possibly be called. For instance, `process_after_download()`  is related to a request sent to `self.message_box.download_sender`.

@albinsuresh  I feels that this will be soon an issue - not only for this actor but for all the complex ones, e.g. the c8y mapper. One needs to find a way to document the actor message flows.

Reply:
We could start with some code comments on the `run()` method.

Reply:
Added some comments.